tes/utils.py

```diff
@@ -27,14 +27,20 @@ def __init__(self, *args, **kwargs):
 
 
 def unmarshal(j: Any, o: Type, convert_camel_case=True) -> Any:
+    m: Any = None
     if isinstance(j, str):
-        m = json.loads(j)
-    elif isinstance(j, dict):
-        m = j
+        try:
+            m = json.loads(j)
+        except json.decoder.JSONDecodeError:
+            pass
     elif j is None:
         return None
     else:
-        raise TypeError("j must be a str, a dict or None")
+        m = j
+
+    if not isinstance(m, dict):
+        raise TypeError("j must be a dictionary, a JSON string evaluation to "
+                        "a dictionary, or None")
```

Comment:
Previously the behavior of `unmarshal()` was not well defined for strings that could be parsed to JSON but that do _not_ end up being dictionaries (`list`, `int`, `float`, `bool`, `None`). And for those strings that could not be parsed, a JSON decode error would be implicitly raised by the code, which I thought was a little inconsistent.

So here I am ignoring decode parsers but then check that we are really dealing with a dictionary later on, and if not, give a single specific message of what input is expected here (dict, JSON string evaluating to dict, or `None`). I find it a bit easier to read as well.